### DISCUSSION TITLE
Take multiple hl7 files and convert them to a single batch file

### DISCUSSION THREAD

**Discussion by MirthNube** [nextgenhealthcare/connect#6004] [General]
Hi All, I am looking for some guidance on how to approach this , I am new to mirth and JS programming in general.

 I need to take multiple incoming HL7 messages and convert them to a single batch file. Does anyone have any suggestions for batching the messages into a single file ?

**Comment by ab-mg-23** [comment]
I recommend starting with doing a quick read of the [Aggregator](https://www.enterpriseintegrationpatterns.com/patterns/messaging/Aggregator.html) pattern. Some questions to consider before starting are: 

- What constitutes a complete set?
- How big can a complete set be?
- What can I use as the correlation key for related messages?
- What do I do if after X time I don't have complete set of messages for some correlation key?
- How can/do I signal to the sending system there's a "missing" message from a set?

**Reply by MirthNube** [reply]
Hello , I have come up with the following ... I am not getting the messages into a batch file they are still being written individually.

**Reply by MirthNube** [reply]
// Initialize global variables
var messageBatch = [];
var batchSize = 115;

// Process incoming messages
function handleMSHMessage(message) {
    // Add the message to the batch
    messageBatch.push(message);

    // Check if the batch size has been reached
    if (messageBatch.length >= batchSize) {
        // Combine messages and write to a file
        writeBatchToFile();
        
        // Reset the batch
        messageBatch = [];
    }
}

// Function to combine and write batch to a file
function writeBatchToFile() {
    var combinedMessage = "";

    for (var i = 0; i < messageBatch.length; i++) {
        combinedMessage += messageBatch[i].getMessage();
    }

    // Write the batch to a file
    FileUtil.writeFile("/path/to/output/file.hl7", combinedMessage);
}

// Process incoming MSH messages
handleMSHMessage(msg);

**Reply by pacmano1** [reply]
Please describe how you are getting the messages into mirth?  Is this a file reader?  a TCP listener?
